src/storage-controller/src/lib.rs

```diff
@@ -289,28 +290,58 @@ where
         &self.config
     }
 
-    fn collection(
+    fn collection_description(
         &self,
         id: GlobalId,
-    ) -> Result<&CollectionState<Self::Timestamp>, StorageError<Self::Timestamp>> {
-        self.collections
-            .get(&id)
-            .ok_or(StorageError::IdentifierMissing(id))
+    ) -> Result<CollectionDescription<Self::Timestamp>, StorageError<Self::Timestamp>> {
+        let res = self.collection(id)?.description.clone();
+
+        Ok(res)
     }
 
-    fn collection_mut(
-        &mut self,
+    fn collection_metadata(
+        &self,
         id: GlobalId,
-    ) -> Result<&mut CollectionState<Self::Timestamp>, StorageError<Self::Timestamp>> {
-        self.collections
-            .get_mut(&id)
-            .ok_or(StorageError::IdentifierMissing(id))
+    ) -> Result<CollectionMetadata, StorageError<Self::Timestamp>> {
+        let res = self.collection(id)?.collection_metadata.clone();
+
+        Ok(res)
     }
 
-    fn collections(
+    fn collection_frontiers(
         &self,
-    ) -> Box<dyn Iterator<Item = (&GlobalId, &CollectionState<Self::Timestamp>)> + '_> {
-        Box::new(self.collections.iter())
+        ids: Vec<GlobalId>,
+    ) -> Result<Vec<(GlobalId, Antichain<T>, Antichain<T>)>, StorageError<Self::Timestamp>> {
+        let res = ids
+            .into_iter()
+            .map(|id| {
+                self.collections
+                    .get(&id)
+                    .map(|c| {
+                        (
+                            id.clone(),
+                            c.implied_capability.clone(),
+                            c.write_frontier.clone(),
+                        )
+                    })
+                    .ok_or(StorageError::IdentifierMissing(id))
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+
+        Ok(res)
+    }
+
+    fn active_collection_metadatas(&self) -> Vec<(GlobalId, CollectionMetadata)> {
+        self.collections
+            .iter()
+            .filter(|(_id, c)| !c.is_dropped())
+            .map(|(id, c)| (*id, c.collection_metadata.clone()))
+            .collect()
+    }
+
+    fn check_exists(&self, id: GlobalId) -> Result<(), StorageError<Self::Timestamp>> {
+        let _collection = self.collection(id)?;
+        Ok(())
     }
```

Comment:
This will always only return a single error, right? Just having a `collection_exists(&self, GlobalId) -> bool` method might be more ergonomic.

Reply:
I did this to keep close to how things currently work, plus I'd now have to manufacture errors at the callsite, if this returned a boolean.

Reply:
Oh, I thought most callsites translate this into a different error anyway and would become simpler with that change. But that might not be true :)